### PR TITLE
[flutter_plugin_tools] Migrate publish and version checks to NNBD

### PR DESCRIPTION
**Pull request by stuartmorgan**
Migrates publish-check and version-check commands to NNBD.

Reworks the version-check flow so that it's more consistent with the other commands: instead of immediately exiting on failure, it checks all plugins, gathers failures, and summarizes all failures at the end. This ensures that we don't have failures in one package temporarily masked by failures in another, so PRs don't need to go through as many check cycles.

Part of https://github.com/flutter/flutter/issues/81912

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`. See [plugin_tool format])
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[plugin_tool format]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

File: script/tool/lib/src/version_check_command.dart [hunk 1]
```diff
@@ -208,38 +210,65 @@ ${indentation}HTTP response: ${pubVersionFinderResponse.httpResponse.body}
...
+      printError('''
+The following pubspecs failed validaton:
+$indentation${badVersionChangePubspecs.join('\n$indentation')}
+''');
+    }
+    if (mismatchedVersionPlugins.isNotEmpty) {
+      passed = false;
+      printError('''
+The following pubspecs failed validaton:
+$indentation${badVersionChangePubspecs.join('\n$indentation')}
```
Comment (cyanglaz) [inline script/tool/lib/src/version_check_command.dart:242]: This seems need to be a different message? Something like: "The following plugins' versions on CHANGELOG and pubspec.yaml are not matched."

File: script/tool/lib/src/common.dart [hunk 2]
```diff
@@ -146,11 +146,10 @@ bool isLinuxPlugin(FileSystemEntity entity) {
   return pluginSupportsPlatform(kLinux, entity);
 }
 
-/// Throws a [ToolExit] with `exitCode` and log the `errorMessage` in red.
-void printErrorAndExit({required String errorMessage, int exitCode = 1}) {
+/// Prints `errorMessage` in red.
+void printError(String errorMessage) {
```
[the remark this answers is not in the document]
Comment (cyanglaz) [reply script/tool/lib/src/common.dart:150]: Oh I see, you want to loop through the plugins instead of exit early. LGTM

**Review by cyanglaz — CHANGES_REQUESTED** [review]

File: script/tool/lib/src/version_check_command.dart (hunk 1 above)
Comment (stuartmorgan) [reply script/tool/lib/src/version_check_command.dart]: I was printing the wrong list of plugins here too; thanks for catching the copypasta!

**Review by cyanglaz — APPROVED** [review]
LGTM

**Pull request merged**